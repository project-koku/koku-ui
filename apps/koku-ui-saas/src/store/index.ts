export { configureStore } from './store';
export type { RootState } from './rootReducer';
