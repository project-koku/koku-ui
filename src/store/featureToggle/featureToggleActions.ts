import { createAction } from 'typesafe-actions';

export interface FeatureToggleActionMeta {
  isAccountInfoDetailsToggleEnabled?: boolean;
  isAccountInfoEmptyStateToggleEnabled?: boolean;
  isAwsEc2InstancesToggleEnabled?: boolean;
  isChartSkeletonToggleEnabled?: boolean;
  isDebugToggleEnabled?: boolean;
  isDetailsDateRangeToggleEnabled?: boolean;
  isExportsToggleEnabled?: boolean;
  isFinsightsToggleEnabled?: boolean;
  isIbmToggleEnabled?: boolean;
  isOcpCloudGroupBysToggleEnabled?: boolean;
  isProviderEmptyStateToggleEnabled?: boolean;
  isVirtualizationToggleEnabled?: boolean;
}

export const setFeatureToggle = createAction('feature/init_feature_toggle')<FeatureToggleActionMeta>();
export const resetState = createAction('feature/reset_state')();
