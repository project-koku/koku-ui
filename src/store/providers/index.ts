import {
  awsProvidersQuery,
  azureProvidersQuery,
  gcpProvidersQuery,
  ibmProvidersQuery,
  ociProvidersQuery,
  ocpProvidersQuery,
  providersQuery,
} from 'store/providers/providersCommon';

import * as providersActions from './providersActions';
import { stateKey as providersStateKey } from './providersCommon';
import { ProvidersAction, providersReducer, ProvidersState } from './providersReducer';
import * as providersSelectors from './providersSelectors';

export {
  awsProvidersQuery,
  azureProvidersQuery,
  ociProvidersQuery,
  gcpProvidersQuery,
  ibmProvidersQuery,
  ocpProvidersQuery,
  ProvidersAction,
  providersActions,
  providersQuery,
  providersReducer,
  providersSelectors,
  ProvidersState,
  providersStateKey,
};
