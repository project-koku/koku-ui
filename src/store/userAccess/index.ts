import {
  awsUserAccessQuery,
  azureUserAccessQuery,
  costModelUserAccessQuery,
  gcpUserAccessQuery,
  ocpUserAccessQuery,
  userAccessQuery,
} from 'store/userAccess/userAccessCommon';

import * as userAccessActions from './userAccessActions';
import { stateKey as userAccessStateKey } from './userAccessCommon';
import type { UserAccessAction, UserAccessState } from './userAccessReducer';
import { userAccessReducer } from './userAccessReducer';
import * as userAccessSelectors from './userAccessSelectors';

export {
  userAccessQuery,
  awsUserAccessQuery,
  azureUserAccessQuery,
  costModelUserAccessQuery,
  gcpUserAccessQuery,
  ocpUserAccessQuery,
  userAccessActions,
  userAccessReducer,
  userAccessSelectors,
  userAccessStateKey,
};
export type { UserAccessAction, UserAccessState };
