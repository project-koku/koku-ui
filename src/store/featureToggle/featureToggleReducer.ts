import type { ActionType } from 'typesafe-actions';
import { getType } from 'typesafe-actions';

import type { resetState } from './featureToggleActions';
import { setFeatureToggle } from './featureToggleActions';

export type FeatureToggleAction = ActionType<typeof setFeatureToggle | typeof resetState>;

export type FeatureToggleState = Readonly<{
  hasFeatureToggle: boolean;
  isAccountInfoDetailsToggleEnabled: boolean;
  isAccountInfoEmptyStateToggleEnabled: boolean;
  isAwsEc2InstancesToggleEnabled: boolean;
  isChartSkeletonToggleEnabled: boolean;
  isDebugToggleEnabled: boolean;
  isDetailsDateRangeToggleEnabled: boolean;
  isExportsToggleEnabled: boolean;
  isFinsightsToggleEnabled: boolean;
  isIbmToggleEnabled: boolean;
  isOcpCloudGroupBysToggleEnabled: boolean;
  isProviderEmptyStateToggleEnabled: boolean;
  isVirtualizationToggleEnabled: boolean;
}>;

export const defaultState: FeatureToggleState = {
  hasFeatureToggle: false,
  isAccountInfoDetailsToggleEnabled: false,
  isAccountInfoEmptyStateToggleEnabled: false,
  isAwsEc2InstancesToggleEnabled: false,
  isChartSkeletonToggleEnabled: false,
  isDebugToggleEnabled: false,
  isDetailsDateRangeToggleEnabled: false,
  isExportsToggleEnabled: false,
  isFinsightsToggleEnabled: false,
  isIbmToggleEnabled: false,
  isOcpCloudGroupBysToggleEnabled: false,
  isProviderEmptyStateToggleEnabled: false,
  isVirtualizationToggleEnabled: false,
};

export const stateKey = 'FeatureToggle';

export function FeatureToggleReducer(state = defaultState, action: FeatureToggleAction): FeatureToggleState {
  switch (action.type) {
    case getType(setFeatureToggle):
      return {
        ...state,
        hasFeatureToggle: true,
        isAccountInfoDetailsToggleEnabled: action.payload.isAccountInfoDetailsToggleEnabled,
        isAccountInfoEmptyStateToggleEnabled: action.payload.isAccountInfoEmptyStateToggleEnabled,
        isAwsEc2InstancesToggleEnabled: action.payload.isAwsEc2InstancesToggleEnabled,
        isChartSkeletonToggleEnabled: action.payload.isChartSkeletonToggleEnabled,
        isDebugToggleEnabled: action.payload.isDebugToggleEnabled,
        isDetailsDateRangeToggleEnabled: action.payload.isDetailsDateRangeToggleEnabled,
        isExportsToggleEnabled: action.payload.isExportsToggleEnabled,
        isFinsightsToggleEnabled: action.payload.isFinsightsToggleEnabled,
        isIbmToggleEnabled: action.payload.isIbmToggleEnabled,
        isOcpCloudGroupBysToggleEnabled: action.payload.isOcpCloudGroupBysToggleEnabled,
        isProviderEmptyStateToggleEnabled: action.payload.isProviderEmptyStateToggleEnabled,
        isVirtualizationToggleEnabled: action.payload.isVirtualizationToggleEnabled,
      };

    default:
      return state;
  }
}
