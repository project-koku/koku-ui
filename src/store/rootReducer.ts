import { notifications } from '@redhat-cloud-services/frontend-components-notifications';
import { combineReducers } from 'redux';
import { awsCostOverviewReducer, awsCostOverviewStateKey } from 'store/breakdown/costOverview/awsCostOverview';
import { azureCostOverviewReducer, azureCostOverviewStateKey } from 'store/breakdown/costOverview/azureCostOverview';
import { gcpCostOverviewReducer, gcpCostOverviewStateKey } from 'store/breakdown/costOverview/gcpCostOverview';
import { ibmCostOverviewReducer, ibmCostOverviewStateKey } from 'store/breakdown/costOverview/ibmCostOverview';
import { ocpCostOverviewReducer, ocpCostOverviewStateKey } from 'store/breakdown/costOverview/ocpCostOverview';
import { awsHistoricalDataReducer, awsHistoricalDataStateKey } from 'store/breakdown/historicalData/awsHistoricalData';
import {
  azureHistoricalDataReducer,
  azureHistoricalDataStateKey,
} from 'store/breakdown/historicalData/azureHistoricalData';
import { gcpHistoricalDataReducer, gcpHistoricalDataStateKey } from 'store/breakdown/historicalData/gcpHistoricalData';
import { ibmHistoricalDataReducer, ibmHistoricalDataStateKey } from 'store/breakdown/historicalData/ibmHistoricalData';
import { ocpHistoricalDataReducer, ocpHistoricalDataStateKey } from 'store/breakdown/historicalData/ocpHistoricalData';
import { costModelsReducer, costModelsStateKey } from 'store/costModels';
import { awsDashboardReducer, awsDashboardStateKey } from 'store/dashboard/awsDashboard';
import { awsOcpDashboardReducer, awsOcpDashboardStateKey } from 'store/dashboard/awsOcpDashboard';
import { azureDashboardReducer, azureDashboardStateKey } from 'store/dashboard/azureDashboard';
import { azureOcpDashboardReducer, azureOcpDashboardStateKey } from 'store/dashboard/azureOcpDashboard';
import { gcpDashboardReducer, gcpDashboardStateKey } from 'store/dashboard/gcpDashboard';
import { gcpOcpDashboardReducer, gcpOcpDashboardStateKey } from 'store/dashboard/gcpOcpDashboard';
import { ibmDashboardReducer, ibmDashboardStateKey } from 'store/dashboard/ibmDashboard';
import { ocpCloudDashboardReducer, ocpCloudDashboardStateKey } from 'store/dashboard/ocpCloudDashboard';
import { ocpDashboardReducer, ocpDashboardStateKey } from 'store/dashboard/ocpDashboard';
import {
  ocpInfrastructureDashboardReducer,
  ocpInfrastructureDashboardStateKey,
} from 'store/dashboard/ocpInfrastructureDashboard';
import {
  ocpSupplementaryDashboardReducer,
  ocpSupplementaryDashboardStateKey,
} from 'store/dashboard/ocpSupplementaryDashboard';
import { ocpUsageDashboardReducer, ocpUsageDashboardStateKey } from 'store/dashboard/ocpUsageDashboard';
import { exportReducer, exportStateKey } from 'store/exports';
import { forecastReducer, forecastStateKey } from 'store/forecasts';
import { orgReducer, orgStateKey } from 'store/orgs';
import { priceListReducer, priceListStateKey } from 'store/priceList';
import { reportReducer, reportStateKey } from 'store/reports';
import { resourceReducer, resourceStateKey } from 'store/resources';
import { sourcesReducer, sourcesStateKey } from 'store/sourceSettings';
import { tagReducer, tagStateKey } from 'store/tags';
import { StateType } from 'typesafe-actions';

import { metricsReducer, metricsStateKey } from './metrics';
import { providersReducer, providersStateKey } from './providers';
import { rbacReducer, rbacStateKey } from './rbac';
import { uiReducer, uiStateKey } from './ui';
import { userAccessReducer, userAccessStateKey } from './userAccess';

export type RootState = StateType<typeof rootReducer>;

export const rootReducer = combineReducers({
  [awsOcpDashboardStateKey]: awsOcpDashboardReducer,
  [awsCostOverviewStateKey]: awsCostOverviewReducer,
  [awsDashboardStateKey]: awsDashboardReducer,
  [awsHistoricalDataStateKey]: awsHistoricalDataReducer,
  [azureOcpDashboardStateKey]: azureOcpDashboardReducer,
  [azureCostOverviewStateKey]: azureCostOverviewReducer,
  [azureDashboardStateKey]: azureDashboardReducer,
  [azureHistoricalDataStateKey]: azureHistoricalDataReducer,
  [costModelsStateKey]: costModelsReducer,
  [exportStateKey]: exportReducer,
  [gcpCostOverviewStateKey]: gcpCostOverviewReducer,
  [gcpDashboardStateKey]: gcpDashboardReducer,
  [gcpOcpDashboardStateKey]: gcpOcpDashboardReducer,
  [gcpHistoricalDataStateKey]: gcpHistoricalDataReducer,
  [ibmCostOverviewStateKey]: ibmCostOverviewReducer,
  [ibmDashboardStateKey]: ibmDashboardReducer,
  [ibmHistoricalDataStateKey]: ibmHistoricalDataReducer,
  [metricsStateKey]: metricsReducer,
  [ocpCostOverviewStateKey]: ocpCostOverviewReducer,
  [ocpDashboardStateKey]: ocpDashboardReducer,
  [ocpCloudDashboardStateKey]: ocpCloudDashboardReducer,
  [ocpHistoricalDataStateKey]: ocpHistoricalDataReducer,
  [ocpInfrastructureDashboardStateKey]: ocpInfrastructureDashboardReducer,
  [ocpSupplementaryDashboardStateKey]: ocpSupplementaryDashboardReducer,
  [ocpUsageDashboardStateKey]: ocpUsageDashboardReducer,
  [orgStateKey]: orgReducer,
  [priceListStateKey]: priceListReducer,
  [providersStateKey]: providersReducer,
  [rbacStateKey]: rbacReducer,
  [reportStateKey]: reportReducer,
  [forecastStateKey]: forecastReducer,
  [resourceStateKey]: resourceReducer,
  [sourcesStateKey]: sourcesReducer,
  [tagStateKey]: tagReducer,
  [uiStateKey]: uiReducer,
  [userAccessStateKey]: userAccessReducer,
  notifications,
});
