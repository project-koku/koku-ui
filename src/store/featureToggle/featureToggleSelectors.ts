import type { RootState } from 'store/rootReducer';

import { stateKey } from './featureToggleReducer';

export const selectFeatureToggleState = (state: RootState) => state[stateKey];

export const selectHasFeatureToggle = (state: RootState) => selectFeatureToggleState(state).hasFeatureToggle;

export const selectIsAccountInfoDetailsToggleEnabled = (state: RootState) =>
  selectFeatureToggleState(state).isAccountInfoDetailsToggleEnabled;
export const selectIsAccountInfoEmptyStateToggleEnabled = (state: RootState) =>
  selectFeatureToggleState(state).isAccountInfoEmptyStateToggleEnabled;
export const selectIsAwsEc2InstancesToggleEnabled = (state: RootState) =>
  selectFeatureToggleState(state).isAwsEc2InstancesToggleEnabled;
export const selectIsChartSkeletonToggleEnabled = (state: RootState) =>
  selectFeatureToggleState(state).isChartSkeletonToggleEnabled;
export const selectIsDebugToggleEnabled = (state: RootState) => selectFeatureToggleState(state).isDebugToggleEnabled;
export const selectIsDetailsDateRangeToggleEnabled = (state: RootState) =>
  selectFeatureToggleState(state).isDetailsDateRangeToggleEnabled;
export const selectIsExportsToggleEnabled = (state: RootState) =>
  selectFeatureToggleState(state).isExportsToggleEnabled;
export const selectIsFinsightsToggleEnabled = (state: RootState) =>
  selectFeatureToggleState(state).isFinsightsToggleEnabled;
export const selectIsIbmToggleEnabled = (state: RootState) => selectFeatureToggleState(state).isIbmToggleEnabled;
export const selectIsOcpCloudGroupBysToggleEnabled = (state: RootState) =>
  selectFeatureToggleState(state).isOcpCloudGroupBysToggleEnabled;
export const selectIsProviderEmptyStateToggleEnabled = (state: RootState) =>
  selectFeatureToggleState(state).isProviderEmptyStateToggleEnabled;
export const selectIsVirtualizationToggleEnabled = (state: RootState) =>
  selectFeatureToggleState(state).isVirtualizationToggleEnabled;
