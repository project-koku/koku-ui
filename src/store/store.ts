import { notificationsMiddleware } from '@redhat-cloud-services/frontend-components-notifications';
import axios from 'axios';
import { applyMiddleware, compose, createStore, DeepPartial } from 'redux';
import thunk from 'redux-thunk';

import { rootReducer, RootState } from './rootReducer';

// eslint-disable-next-line @typescript-eslint/no-unused-vars
declare global {
  interface Window {
    __REDUX_DEVTOOLS_EXTENSION_COMPOSE__(options: any): typeof compose;
  }
}

const composeEnhancers =
  typeof window === 'object' && window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__
    ? window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__({ serialize: true })
    : compose;

export const middlewares = [thunk, notificationsMiddleware()];

export function configureStore(initialState: DeepPartial<RootState>) {
  const enhancer = composeEnhancers(applyMiddleware(...middlewares));
  const store = createStore(rootReducer, initialState as any, enhancer);

  axios.interceptors.response.use(null, error => {
    return Promise.reject(error);
  });

  return store;
}
