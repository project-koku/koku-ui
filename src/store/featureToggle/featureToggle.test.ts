import { createMockStoreCreator } from 'store/mockStore';

import { FeatureToggleSelectors } from '.';
import * as actions from './featureToggleActions';
import { FeatureToggleReducer, stateKey } from './featureToggleReducer';
import * as selectors from './featureToggleSelectors';

const createUIStore = createMockStoreCreator({
  [stateKey]: FeatureToggleReducer,
});

test('default state', async () => {
  const store = createUIStore();
  expect(selectors.selectFeatureToggleState(store.getState())).toMatchSnapshot();
});

test('Cost breakdown chart feature is enabled', async () => {
  const store = createUIStore();
  store.dispatch(actions.setFeatureToggle({ isCostBreakdownChartToggleEnabled: true }));
  expect(FeatureToggleSelectors.selectIsCostBreakdownChartToggleEnabled(store.getState())).toBe(true);
});

test('Debug feature is enabled', async () => {
  const store = createUIStore();
  store.dispatch(actions.setFeatureToggle({ isDebugToggleEnabled: true }));
  expect(FeatureToggleSelectors.selectIsDebugToggleEnabled(store.getState())).toBe(true);
});

test('Exports feature is enabled', async () => {
  const store = createUIStore();
  store.dispatch(actions.setFeatureToggle({ isExportsToggleEnabled: true }));
  expect(FeatureToggleSelectors.selectIsExportsToggleEnabled(store.getState())).toBe(true);
});

test('FINsights feature is enabled', async () => {
  const store = createUIStore();
  store.dispatch(actions.setFeatureToggle({ isFinsightsToggleEnabled: true }));
  expect(FeatureToggleSelectors.selectIsFinsightsToggleEnabled(store.getState())).toBe(true);
});

test('IBM feature is enabled', async () => {
  const store = createUIStore();
  store.dispatch(actions.setFeatureToggle({ isIbmToggleEnabled: true }));
  expect(FeatureToggleSelectors.selectIsIbmToggleEnabled(store.getState())).toBe(true);
});

