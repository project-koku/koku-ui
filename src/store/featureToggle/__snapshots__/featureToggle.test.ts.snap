// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`default state 1`] = `
{
  "hasFeatureToggle": false,
  "isAccountInfoDetailsToggleEnabled": false,
  "isAccountInfoEmptyStateToggleEnabled": false,
  "isAwsEc2InstancesToggleEnabled": false,
  "isChartSkeletonToggleEnabled": false,
  "isDebugToggleEnabled": false,
  "isDetailsDateRangeToggleEnabled": false,
  "isExportsToggleEnabled": false,
  "isFinsightsToggleEnabled": false,
  "isIbmToggleEnabled": false,
  "isOcpCloudGroupBysToggleEnabled": false,
  "isProviderEmptyStateToggleEnabled": false,
  "isVirtualizationToggleEnabled": false,
}
`;
